c7n/resources/s3.py

```diff
@@ -364,6 +364,74 @@ def process_bucket(self, b):
             return b
 
 
+@filters.register('access')
+class AccessFilter(Filter):
+    """Filters for all S3 buckets that have access allowed/denied to a role
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: s3-perm-check
+                resource: s3
+                filters:
+                  - type: access
+                    value: denied
+                    policy_source: arn:::
+    """
+
+    schema = type_schema('access',
+        policy_source={'type': 'string'},
+        caller={'type': 'string', 'required': False},
+        actions={'type': 'array', 'items': {'type': 'string'}},
+        value={'type': 'string', 'enum': ['allowed', 'denied']})
+
+    permissions = ('s3:ListBucket',)
+
+    def process(self, buckets, event=None):
+        with self.executor_factory(max_workers=5) as w:
+            results = w.map(self.process_bucket, buckets)
+            results = filter(None, list(results))
+            return results
+
+    def process_bucket(self, b):
+        session = local_session(self.manager.session_factory)
+        iam = session.client('iam')
+        s3 = session.client('s3')
+
+        policy_source = self.data['policy_source']
+
+        context_keys_extra = {}
+        simulate_extra = {}
+        try:
+            resource_policy = s3.get_bucket_policy(
+                Bucket=b['Name'])['Policy']
+        except ClientError as err:
+            if err.response['Error']['Code'] != 'NoSuchBucketPolicy':
+                raise
+            resource_policy = None
+        else:
+            context_keys_extra['PolicyInputList'] = [resource_policy]
+            simulate_extra['ResourcePolicy'] = resource_policy
+
+        response = iam.get_context_keys_for_principal_policy(
+            PolicySourceArn=policy_source, **context_keys_extra)
+        if response.get('ContextKeyNames', []):
+            raise NotImplementedError("Sorry, we can't handle context keys.")
+
+        response = iam.simulate_principal_policy(
+            PolicySourceArn=policy_source,
+            CallerArn=self.data.get('caller', policy_source),
+            ActionNames=self.data.get('actions', ['s3:*']),
+            ResourceArns=['arn:aws:s3:::{}/*'.format(b['Name'])],
+            **simulate_extra)
+        decision = response['EvaluationResults'][0]['EvalDecision']
+        if decision == self.data['value']:
```

Comment:
There's some additional information from the simulate decision i think is worth annotating onto the resource.